Adds a dev container to help get started with GitHub Codespaces

samruddhikhandale opened the pull request:
Hi 👋 

I am Samruddhi, a Software Engineer for GitHub, working with the [GitHub Codespaces](https://github.com/features/codespaces) team, and a maintainer for the [dev containers](https://github.com/devcontainers) org 👩‍💻

I would like to contribute a dev container to this repo and update documentation to improve the GitHub Codespaces experience.

![image](https://user-images.githubusercontent.com/24955913/221273495-68ab9044-f2a4-406e-9f98-fa28c05dada0.png)

You can test these changes with Codespaces by simply clicking on the `Code` icon on top right side of this PR ☝️ 

ℹ️ Shown interest for Codespaces in https://github.com/astropy/astropy-project/issues/255#issuecomment-1361603217

Thanks,
Samruddhi

pllim commented:
Wow, thank you! I think those interested in Codespaces in the Astropy community would be very grateful for this PR. 😸

taldcroft commented:
@samruddhikhandale - this is really exciting! I just tried it out and it brought up a web VS code but with some ominous warnings/errors along the way.

In the terminal:
```
This codespace is currently running in recovery mode due to a container error.
1. Use Cmd/Ctrl + Shift + P -> "Codespaces: View Creation Log" to see full logs
2. Update your devcontainer configuration as needed
3. Use Cmd/Ctrl + Shift + P -> "Codespaces: Rebuild Container" to retry
4. For help, read more about custom configuration: https://aka.ms/ghcs-custom-configuration
@taldcroft ➜ /workspaces/astropy (samruddhikhandale/add-dev-container) $ 
```

creation.log:
```
=================================================================================
2023-03-01 22:15:58.886Z: Configuration starting...
2023-03-01 22:15:58.944Z: Cloning...
2023-03-01 22:15:58.985Z: $ git -C "/var/lib/docker/codespacemount/workspace" clone --branch samruddhikhandale/add-dev-container --depth 1 https://github.com/samruddhikhandale/astropy "/var/lib/docker/codespacemount/workspace/astropy"
2023-03-01 22:15:59.003Z: Cloning into '/var/lib/docker/codespacemount/workspace/astropy'...
2023-03-01 22:16:03.972Z: git process exited with exit code 0
2023-03-01 22:16:04.019Z: $ git -C "/var/lib/docker/codespacemount/workspace/astropy" config --local remote.origin.fetch +refs/heads/*:refs/remotes/origin/*
2023-03-01 22:16:04.022Z: git process exited with exit code 0

=================================================================================
2023-03-01 22:16:04.587Z: Creating container...
2023-03-01 22:16:04.597Z: $ devcontainer up --id-label Type=codespaces --workspace-folder /var/lib/docker/codespacemount/workspace/astropy --mount type=bind,source=/.codespaces/agent/mount/cache,target=/vscode --user-data-folder /var/lib/docker/codespacemount/.persistedshare --container-data-folder .vscode-remote/data/Machine --container-system-data-folder /var/vscode-remote --log-level trace --log-format json --update-remote-user-uid-default never --mount-workspace-git-root false --skip-non-blocking-commands --skip-post-create --config "/var/lib/docker/codespacemount/workspace/astropy/.devcontainer/devcontainer.json" --override-config /root/.codespaces/shared/merged_devcontainer.json --default-user-env-probe loginInteractiveShell --container-session-data-folder /workspaces/.codespaces/.persistedshare/devcontainers-cli/cache
2023-03-01 22:16:04.815Z: @devcontainers/cli 0.28.0. Node.js v14.21.3. linux 5.4.0-1103-azure x64.
2023-03-01 22:16:07.881Z: Error: Command failed: docker inspect --type image mcr.microsoft.com/devcontainers/miniconda:0-3
2023-03-01 22:16:07.885Z:     at fie (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:1915:1355)
2023-03-01 22:16:07.887Z:     at P7 (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:1915:1291)
2023-03-01 22:16:07.888Z:     at processTicksAndRejections (internal/process/task_queues.js:95:5)
2023-03-01 22:16:07.889Z:     at async Fie (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:1921:2093)
2023-03-01 22:16:07.890Z:     at async $f (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:1921:3239)
2023-03-01 22:16:07.891Z:     at async eoe (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:2045:17247)
2023-03-01 22:16:07.903Z:     at async Qse (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:2045:16993)
2023-03-01 22:16:07.905Z: {"outcome":"error","message":"Command failed: docker inspect --type image mcr.microsoft.com/devcontainers/miniconda:0-3","description":"An error occurred setting up the container."}
2023-03-01 22:16:07.932Z: devcontainer process exited with exit code 1

====================================== ERROR ====================================
2023-03-01 22:16:07.934Z: Failed to create container.
=================================================================================
2023-03-01 22:16:07.935Z: Error: Command failed: docker inspect --type image mcr.microsoft.com/devcontainers/miniconda:0-3
2023-03-01 22:16:07.936Z: Error Code: 1302

====================================== ERROR ====================================
2023-03-01 22:16:07.937Z: Container creation failed.
=================================================================================
2023-03-01 22:16:07.945Z: 

===================================== WARNING ===================================
2023-03-01 22:16:07.947Z: Creating recovery container.
=================================================================================

=================================================================================
2023-03-01 22:16:56.334Z: Creating container...
2023-03-01 22:16:56.337Z: $ devcontainer up --id-label Type=codespaces --workspace-folder /var/lib/docker/codespacemount/workspace/astropy --mount type=bind,source=/.codespaces/agent/mount/cache,target=/vscode --user-data-folder /var/lib/docker/codespacemount/.persistedshare --container-data-folder .vscode-remote/data/Machine --container-system-data-folder /var/vscode-remote --log-level trace --log-format json --update-remote-user-uid-default never --mount-workspace-git-root false --skip-non-blocking-commands --skip-post-create --config "/var/lib/docker/codespacemount/workspace/astropy/.devcontainer/devcontainer.json" --override-config /root/.codespaces/shared/merged_devcontainer.json --default-user-env-probe loginInteractiveShell --container-session-data-folder /workspaces/.codespaces/.persistedshare/devcontainers-cli/cache
2023-03-01 22:16:56.546Z: @devcontainers/cli 0.28.0. Node.js v14.21.3. linux 5.4.0-1103-azure x64.
2023-03-01 22:17:11.283Z: $alpine -c echo Container started
2023-03-01 22:17:11.720Z: Container started
2023-03-01 22:17:12.062Z: Outcome: success User: vscode WorkspaceFolder: /workspaces/astropy
2023-03-01 22:17:12.077Z: devcontainer process exited with exit code 0

=================================================================================
2023-03-01 22:17:14.156Z: Running blocking commands...
2023-03-01 22:17:14.164Z: $ devcontainer up --id-label Type=codespaces --workspace-folder /var/lib/docker/codespacemount/workspace/astropy --mount type=bind,source=/.codespaces/agent/mount/cache,target=/vscode --user-data-folder /var/lib/docker/codespacemount/.persistedshare --container-data-folder .vscode-remote/data/Machine --container-system-data-folder /var/vscode-remote --log-level trace --log-format json --update-remote-user-uid-default never --mount-workspace-git-root false --skip-non-blocking-commands --expect-existing-container --config "/var/lib/docker/codespacemount/workspace/astropy/.devcontainer/devcontainer.json" --override-config /root/.codespaces/shared/merged_devcontainer.json --default-user-env-probe loginInteractiveShell --container-session-data-folder /workspaces/.codespaces/.persistedshare/devcontainers-cli/cache
2023-03-01 22:17:14.383Z: @devcontainers/cli 0.28.0. Node.js v14.21.3. linux 5.4.0-1103-azure x64.
2023-03-01 22:17:14.678Z: Outcome: success User: vscode WorkspaceFolder: /workspaces/astropy
2023-03-01 22:17:14.781Z: devcontainer process exited with exit code 0

=================================================================================
2023-03-01 22:17:14.904Z: Configuring codespace...

=================================================================================
2023-03-01 22:17:14.985Z: Finished configuring codespace.
```

taldcroft commented:
OK, I followed the suggestion for `Codespaces: Rebuild workspace` and it was successful this time!

I'm trying to run tests now using `tox -e py310-test`, so far so good.

samruddhikhandale commented:
Hah! Me and my co-worker are able to successfully create a codespace.

> docker inspect --type image mcr.microsoft.com/devcontainers/miniconda:0-3
For some reason `docker inspect` seems to be flakey, @taldcroft do you mind retrying?

taldcroft commented:
@samruddhikhandale - yes, see the comment above. I did retry and it worked 2nd time and the unit tests are successfully running now.

samruddhikhandale commented:
Nice, thanks for confirming!

taldcroft commented:
And best of all I love my dev container name: `https://taldcroft-studious-space-goggles-....dev`. Is that permanent?

taldcroft commented:
Probably not related to the environment, but sphinx gave two doctest errors. Otherwise looking great:
```
========================================================================== short test summary info ===========================================================================
FAILED ../../.tox/py310-test/lib/python3.10/site-packages/astropy/utils/introspection.py::astropy.utils.introspection.minversion
FAILED ../../docs/timeseries/io.rst::io.rst
=================================================== 2 failed, 21666 passed, 3603 skipped, 176 xfailed in 409.17s (0:06:49) ===================================================
```

taldcroft commented:
I just tried to do something I would want in a dev environment, namely install `ipython`. What I tried didn't quite work:
```
@taldcroft ➜ /workspaces/astropy (samruddhikhandale/add-dev-container) $ conda install ipython
Collecting package metadata (current_repodata.json): done
Solving environment: - 
The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/linux-64::requests==2.28.1=py310h06a4308_0
  - defaults/linux-64::conda==22.11.1=py310h06a4308_4
  - defaults/linux-64::conda-content-trust==0.1.3=py310h06a4308_0
  - defaults/noarch::pyopenssl==22.0.0=pyhd3eb1b0_0
  - defaults/linux-64::urllib3==1.26.13=py310h06a4308_0
done


==> WARNING: A newer version of conda exists. <==
  current version: 22.11.1
  latest version: 23.1.0

Please update conda by running

    $ conda update -n base -c defaults conda

Or to minimize the number of packages updated during conda update use

     conda install conda=23.1.0



## Package Plan ##

  environment location: /opt/conda

  added / updated specs:
    - ipython


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    asttokens-2.0.5            |     pyhd3eb1b0_0          20 KB
    backcall-0.2.0             |     pyhd3eb1b0_0          13 KB
    ca-certificates-2023.01.10 |       h06a4308_0         120 KB
    cryptography-37.0.1        |  py310h9ce1e76_0         1.5 MB
    decorator-5.1.1            |     pyhd3eb1b0_0          12 KB
    executing-0.8.3            |     pyhd3eb1b0_0          18 KB
    ipython-8.10.0             |  py310h06a4308_0         1.1 MB
    jedi-0.18.1                |  py310h06a4308_1         988 KB
    matplotlib-inline-0.1.6    |  py310h06a4308_0          16 KB
    openssl-1.1.1t             |       h7f8727e_0         3.7 MB
    parso-0.8.3                |     pyhd3eb1b0_0          70 KB
    pexpect-4.8.0              |     pyhd3eb1b0_3          53 KB
    pickleshare-0.7.5          |  pyhd3eb1b0_1003          13 KB
    prompt-toolkit-3.0.36      |  py310h06a4308_0         592 KB
    ptyprocess-0.7.0           |     pyhd3eb1b0_2          17 KB
    pure_eval-0.2.2            |     pyhd3eb1b0_0          14 KB
    pygments-2.11.2            |     pyhd3eb1b0_0         759 KB
    stack_data-0.2.0           |     pyhd3eb1b0_0          22 KB
    traitlets-5.7.1            |  py310h06a4308_0         203 KB
    wcwidth-0.2.5              |     pyhd3eb1b0_0          26 KB
    ------------------------------------------------------------
                                           Total:         9.2 MB

The following NEW packages will be INSTALLED:

  asttokens          pkgs/main/noarch::asttokens-2.0.5-pyhd3eb1b0_0 
  backcall           pkgs/main/noarch::backcall-0.2.0-pyhd3eb1b0_0 
  cryptography       pkgs/main/linux-64::cryptography-37.0.1-py310h9ce1e76_0 
  decorator          pkgs/main/noarch::decorator-5.1.1-pyhd3eb1b0_0 
  executing          pkgs/main/noarch::executing-0.8.3-pyhd3eb1b0_0 
  ipython            pkgs/main/linux-64::ipython-8.10.0-py310h06a4308_0 
  jedi               pkgs/main/linux-64::jedi-0.18.1-py310h06a4308_1 
  matplotlib-inline  pkgs/main/linux-64::matplotlib-inline-0.1.6-py310h06a4308_0 
  parso              pkgs/main/noarch::parso-0.8.3-pyhd3eb1b0_0 
  pexpect            pkgs/main/noarch::pexpect-4.8.0-pyhd3eb1b0_3 
  pickleshare        pkgs/main/noarch::pickleshare-0.7.5-pyhd3eb1b0_1003 
  prompt-toolkit     pkgs/main/linux-64::prompt-toolkit-3.0.36-py310h06a4308_0 
  ptyprocess         pkgs/main/noarch::ptyprocess-0.7.0-pyhd3eb1b0_2 
  pure_eval          pkgs/main/noarch::pure_eval-0.2.2-pyhd3eb1b0_0 
  pygments           pkgs/main/noarch::pygments-2.11.2-pyhd3eb1b0_0 
  stack_data         pkgs/main/noarch::stack_data-0.2.0-pyhd3eb1b0_0 
  traitlets          pkgs/main/linux-64::traitlets-5.7.1-py310h06a4308_0 
  wcwidth            pkgs/main/noarch::wcwidth-0.2.5-pyhd3eb1b0_0 

The following packages will be UPDATED:

  ca-certificates                     2022.10.11-h06a4308_0 --> 2023.01.10-h06a4308_0 
  openssl                                 1.1.1s-h7f8727e_0 --> 1.1.1t-h7f8727e_0 


Proceed ([y]/n)? y


Downloading and Extracting Packages
                                                                                                                                                             
Preparing transaction: done                                                                                                                                  
Verifying transaction: done                                                                                                                                  
Executing transaction: - WARNING conda.gateways.disk.delete:unlink_or_rename_to_trash(140): Could not remove or rename /opt/conda/lib/python3.10/site-packages/pygments/__init__.py.  Please remove this file manually (you may need to reboot to free file handles)                                                      
| WARNING conda.gateways.disk.delete:unlink_or_rename_to_trash(140): Could not remove or rename /opt/conda/lib/python3.10/site-packages/cryptography/__about__.py.  Please remove this file manually (you may need to reboot to free file handles)                                                                        
done                                                                                                                                                         
ERROR conda.core.link:_execute(740): An error occurred while installing package 'defaults::pygments-2.11.2-pyhd3eb1b0_0'.                                    
Rolling back transaction: done                                                                                                                               
                                                                                                                                                             
[Errno 13] Permission denied: '/opt/conda/lib/python3.10/site-packages/pygments/__init__.py'                                                                 
()                                                                                                                                                           
```

samruddhikhandale commented:
> And best of all I love my dev container name: `https://taldcroft-studious-space-goggles-....dev`. Is that permanent?

Yes, for the lifetime of that particular codespace.
Every new codespace would have a different name!

samruddhikhandale commented:
https://github.com/astropy/astropy/pull/14481/commits/f27cb8fe53a57de0d7677483076aece872f984e7 adds `ipython` to `requirements.txt`

@taldcroft do you mind syncing your codespace with the latest changes and rebuilding the codespace?

![image](https://user-images.githubusercontent.com/24955913/222285985-a93599f3-7eee-4900-8040-683b0b9aaf00.png)

In the meanwhile, trying to figure why `conda install ..` could have failed.

taldcroft commented:
The new build has `ipython` which apparently works. One new twist, this odd warning about a shallow workspace. I installed astropy as an editable build to reflect how I usually develop/test.
```
@taldcroft ➜ /workspaces/astropy (samruddhikhandale/add-dev-container) $ ipython
Python 3.10.8 (main, Nov 24 2022, 14:13:03) [GCC 11.2.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.11.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import astropy
/opt/conda/lib/python3.10/site-packages/setuptools_scm/git.py:135: UserWarning: "/workspaces/astropy" is shallow and may cause errors
  warnings.warn(f'"{wd.path}" is shallow and may cause errors')
```

samruddhikhandale commented:
> I installed astropy as an editable build to reflect how I usually develop/test.

@taldcroft Can you help me with repro steps?

I tried the following

`$ pip install -e .` followed with 👇 

![image](https://user-images.githubusercontent.com/24955913/222289266-ecdca23a-f771-4e5d-a85d-327abb94d43a.png)

Note: My codespace is built with https://github.com/astropy/astropy/pull/14481/commits/f0357d238ac4938dd7b725c5d0897dbfd6f62ecf changes, not sure if relevant.

taldcroft commented:
@samruddhikhandale - what  do you get for the astropy version? 
```
In [1]: import astropy
/opt/conda/lib/python3.10/site-packages/setuptools_scm/git.py:135: UserWarning: "/workspaces/astropy" is shallow and may cause errors
  warnings.warn(f'"{wd.path}" is shallow and may cause errors')

In [2]: astropy.__version__
Out[2]: '0.1.dev3+gf0357d2'
```
That warning is coming from `setuptools_scm` and now I see that the version is wrong.

taldcroft commented:
BTW I did rebuild with the latest commits. VS code very nicely indicated that there were updates to the PR and that I need to rebuild the codespace. All this integration is really amazing.

samruddhikhandale commented:
> what do you get for the astropy version?

@taldcroft here you go 👇 

```
In [2]: astropy.__version__
Out[2]: '5.3.dev661+gf0357d238'
```

samruddhikhandale commented:
These files requires it to add `setuptools_scm` >= 6.2

![image](https://user-images.githubusercontent.com/24955913/222294603-97073369-ea66-45f6-a8b6-5936f92e36c5.png)

In my codespace, I have `v7.1.0` installed. I did not execute `tox -e py310-test`, could that command change the version?

taldcroft commented:
@samruddhikhandale  - the obvious benefit of this codespace is for occasional contributors so they don't need to do much setup to get going. Apart from that one oddity this is basically meeting that goal and really amazing.

Now I'm wondering about experienced project developers like me. I have a bunch of VS code settings locally and extensions like `black` and `ruff` and so on. Is there a way to sync my local profile to the codespace so I have a consistent dev environment? I basically just work on a single laptop so I don't have any experience with syncing my profile settings.

samruddhikhandale commented:
@taldcroft You can turn on `Settings Sync` which will sync your VS Code settings to your codepsaces.

![image](https://user-images.githubusercontent.com/24955913/222295868-6c28dbf6-2f51-4fb7-9f33-6a39dcbb8f5c.png)

https://docs.github.com/en/codespaces/getting-started/quickstart#about-settings-sync

Also, if you think contributors could benefit from the local extensions you have, then I'm happy to add it to the dev config. Let me know the list, and I'd update this PR 😄

taldcroft commented:
@samruddhikhandale - is there a possibility of any persistence of my original error that is messing things up? Any way for me to start totally from scratch?

I just tried a full rebuild but no joy. Still the same problem and I cannot do something like `conda install black`.

<img width="496" alt="image" src="https://user-images.githubusercontent.com/348089/222296045-bdbae841-690a-453c-85e5-82389618e20c.png">

I noticed after the full rebuild that it still had my old user files, e.g. `.tox/py310-test` was already there.

samruddhikhandale commented:
A `full rebuild` should help, it recreates the entire dev container but keeps the old files intact.
I just tried `conda install black` which was successful , https://github.com/astropy/astropy/pull/14481/commits/f0357d238ac4938dd7b725c5d0897dbfd6f62ecf fixed things for me after a rebuild.

If rebuild din't work, I wonder if creating a new codespace fixes the error for you?

taldcroft commented:
Woohoo, just making a completely new codespace fixed the problem!! `studious-space-goggles` was corrupted somehow.

taldcroft commented:
So now about other dependencies and so forth... for me at least I'd like to see all the optional dependencies, which I believe are captured here:
https://github.com/astropy/astropy/blob/09e54670e4a46ed510e32d8206e4853920684952/setup.cfg#L76

Not on that list but useful is `pre-commit`.

For VS code extensions by default:
- Black formatter
- Ruff
- Rewrap (not so common but I find this super useful)

taldcroft commented:
@samruddhikhandale - I tried to switch to a new branch to test that workflow, but then VS code noticed the change and wanted to rebuild the codespace. So I'm assuming this PR needs to be merged to main before I could play around with making a new PR on astropy?

samruddhikhandale commented:
@taldcroft Addressed the comments, let me know if I may have missed something!

> I tried to switch to a new branch to test that workflow, but then VS code noticed the change and wanted to rebuild the codespace. So I'm assuming this PR needs to be merged to main before I could play around with making a new PR on astropy?

We'd need to merge these changes to `main` for other people to benefit from it. However, you can ignore the`Rebuilding` prompt. That way you can still have a dev container with these PR changes 😅

samruddhikhandale commented:
Sounds wonderful, thank you!

> Assuming we eventually get some contributors taking advantage of this, and then there is some problem or question we can't answer, is there a forum for codespaces support?

Yep, https://github.com/orgs/community/discussions/categories/codespaces
Updated the README with this URL.

maxnoe commented on .devcontainer/requirements.txt:
```diff
@@ -0,0 +1,38 @@
+asdf>=2.10.0
```
Is it really needed to duplicate the requirements here?

Does codespaces not support running `pip install -e[all]`?

dr-rodriguez replied:
I agree, this can be a postCreateCommand to avoid duplicate requirements that could get out of date.

taldcroft reviewed (approved):
This looks great to me! As noted in the discussion I tested this and after an initial hiccup everything worked as expected. This included:
- Build astropy as an editable install.
- Run pytest locally.
- Run `tox -e py310-test` successfully except two doctest tests.
- Conda and pip install packages to the base environment.
- Sync my VS code settings.
- Close the window and successfully come back to the codespace 10 hours later.

falkben commented:
When I opened the codespace I got the following error in the integrated terminal:

```
@falkben ➜ /workspaces/astropy (samruddhikhandale/add-dev-container) $ conda activate base

CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.
To initialize your shell, run

    $ conda init <SHELL_NAME>

Currently supported shells are:
  - bash
  - fish
  - tcsh
  - xonsh
  - zsh
  - powershell

See 'conda init --help' for more information and options.

IMPORTANT: You may need to close and restart your shell after running 'conda init'.


@falkben ➜ /workspaces/astropy (samruddhikhandale/add-dev-container) $ 
```

The following command seems to indicate the conda environment is present:

```
@falkben ➜ /workspaces/astropy (samruddhikhandale/add-dev-container) $ conda env list
# conda environments:
#
base                     /opt/conda
```

dr-rodriguez commented:
I also wonder: why use conda at all if all you are doing is running pip? You can have the .devcontainer just have a python 3.10 environment to start from, see for example: https://github.com/dr-rodriguez/AstrodbKit2/blob/main/.devcontainer/devcontainer.json

dr-rodriguez commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,16 @@
+{
```
The following should replicate the same environment but without the extra Dockerfile, conda, and the need to duplicate the requirements.txt file:
```
{
    "name": "Python 3",
	"image": "mcr.microsoft.com/devcontainers/python:3.10",
    "customizations": {
        "vscode":{
            "extensions": [
                "ms-python.python",
                "ms-python.black-formatter",
                "charliermarsh.ruff",
                "stkb.rewrap"
            ]
        }
    },
    "postCreateCommand": "pip install tox && pip install -e .[all]"
}
```
It still allows you to run `tox -e py310-test` to verify code changes work.

samruddhikhandale replied:
Updated dev container! 👍

taldcroft replied:
@dr-rodriguez - unfortunately that new specification is not working nicely for me like the previous conda-based one was. When I come up in the terminal the path is not right:
```
@taldcroft ➜ /workspaces/astropy (samruddhikhandale/add-dev-container) $ which python
/usr/local/bin/python
@taldcroft ➜ /workspaces/astropy (samruddhikhandale/add-dev-container) $ which pip
/usr/local/bin/pip
```
I preferred the conda environment because that makes it easy for me to create a new environment with a different Python version. It happens sometimes that I need to debug something that only occurs with a particular Python version and conda is my go-to for that.

What I normally do is use conda to create the base environment, then make and activate an `astropy` conda env, then use `pip install -e .[all]` to get all the dependencies and astropy. That gives the best of both worlds IMO.

taldcroft replied:
Also with this change something is broken again with `setuptools_scm`:
```
@taldcroft ➜ /workspaces/astropy (samruddhikhandale/add-dev-container) $ which ipython
/home/vscode/.local/bin/ipython
@taldcroft ➜ /workspaces/astropy (samruddhikhandale/add-dev-container) $ ipython
Python 3.10.10 (main, Feb 11 2023, 02:40:33) [GCC 10.2.1 20210110]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.11.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import astropy

In [2]: astropy.__version__
Out[2]: '0.1.dev1+gb06e6cd'
```

samruddhikhandale commented:
> I also wonder: why use conda at all if all you are doing is running pip? You can have the .devcontainer just have a python 3.10 environment to start from, see for example: https://github.com/dr-rodriguez/AstrodbKit2/blob/main/.devcontainer/devcontainer.json

From the [docs](https://docs.astropy.org/en/latest/development/workflow/virtual_pythons.html#virtual-envs), I assumed `conda` is needed for creating virtiual env. Hence, thought of using the `miniconda` image. If that's not the case, then I can switch to `python` 👍 

Will update this PR in a bit.

samruddhikhandale commented:
Refactored dev container with https://github.com/astropy/astropy/pull/14481#discussion_r1123614113 suggesstion.

Note: New dev container no more contains `conda`, and `make`.

taldcroft commented:
> why use conda at all if all you are doing is running pip?

As mentioned in a code comment, using conda is useful because it makes it easy to create an environment with a different Python version for testing. I do this frequently.

dr-rodriguez commented:
> > why use conda at all if all you are doing is running pip?
> 
> As mentioned in a code comment, using conda is useful because it makes it easy to create an environment with a different Python version for testing. I do this frequently.

As I understand it, a Codespace is an environment so when we need a different python version we would edit the devcontainer.json and rebuild the container. It's a different workflow, but should have the same outcome. That said, it does take a little longer and if this is breaking for you then maybe something else is missing or is incompatible. I have no strong preference for/against conda, so if most users would prefer conda then maybe it's fine to use that.

taldcroft commented:
> when we need a different python version we would edit the devcontainer.json and rebuild the container.

Swapping to a different supported version of Python should be a quick thing that is possible within the same container. The reason being that I will have some test code or patch and I want to be able to run the same code locally in different environments.

taldcroft commented:
> if this is breaking for you ...

Indeed it is.

BTW about the `conda activate base` issue, that was easily solved for me by following the instructions in the conda message to run `conda init bash` and restarting the terminal. Perhaps this `conda init bash` could be put into the setup?

samruddhikhandale commented:
👋  Pushed a change with the latest iteration of feedbacks! Let me know if the dev container looks good.

<img width="1192" alt="image" src="https://user-images.githubusercontent.com/24955913/222589974-98c5412f-7f2c-492d-b53a-f4ef9e6895d5.png">

Executed `tox -e py310-test` successfully except two doctest. 

<img width="1303" alt="image" src="https://user-images.githubusercontent.com/24955913/222590803-9dffa6f3-5d10-4326-953f-e79f3d59a71f.png">

taldcroft commented:
@samruddhikhandale - thanks, this is looking good again!

Now the Terminal correctly activates the `base` environment the first time, yay. 

I've noticed each time I make a brand new codespace that I get something like below. It is easily fixed by clicking the button and reloading the editor, but I wonder if there is some configuration issue the is preventing the `Python` extension from loading initially.

<img width="452" alt="image" src="https://user-images.githubusercontent.com/348089/222711708-89bdf1e4-9244-4634-9c94-5397933274be.png">

Some changes that I think are still needed:

- Change `pip install -e .[all]` to `pip install -e .[all,test_all,docs]`. The latter two are needed to run tests with `pytest` and build the docs.
- Before doing the`pip install` there needs to be something like `git fetch --tag`. 
  - This was the source of my version problems in the editable install because the initial checkout of the repo had no tags, so `setuptools_scm` had no reference point.
  - I was a little surprised that `git fetch --tag` took quite a while (at least 30 seconds) with a lot of processing, so I'm not sure what's going on behind the scenes there, but that definitely fixed the version issue I was seeing.

taldcroft commented:
I note that building the docs seems to be failing. I got this twice in a row:
```
cd docs
make html
...
```
<img width="716" alt="image" src="https://user-images.githubusercontent.com/348089/222714917-bd87bd9c-a7b2-4634-a656-ff399356bfa4.png">

This is lower priority so I would not consider this a blocker for merging, but I'm curious why it does not work.

I'm currently trying `tox -e build_docs`, which is the nominal official way to build the docs.

taldcroft commented:
Ah OK, now I'm seeing that building the docs is a bit CPU expensive:
<img width="459" alt="image" src="https://user-images.githubusercontent.com/348089/222715918-b6669890-a47a-4654-8cf5-a10a69fe3d02.png">

taldcroft commented:
The `tox -e build_docs` succeeded. Then I expected I could look at the rendered output in VS code live preview, but it didn't work:
<img width="459" alt="image" src="https://user-images.githubusercontent.com/348089/222718239-5e93921b-00ef-4d72-a9d2-dece0b1aac3a.png">

But then some message came up about seeing the local files via port 3000 and I was able to view view a URL like: ```https://taldcroft-...-3000.preview.app.github.dev/docs/_build/html/`. Interesting.

pllim commented on README.rst:
```diff
@@ -33,6 +33,17 @@ page. A `summary of contribution guidelines <CONTRIBUTING.md>`_ can also be
 used as a quick reference when you are ready to start writing or validating
 code for submission.
 
+Getting started with GitHub Codespaces
+----------------------
```
```suggestion
Getting started with GitHub Codespaces
--------------------------------------
```

eteq commented on README.rst:
```diff
@@ -33,6 +33,17 @@ page. A `summary of contribution guidelines <CONTRIBUTING.md>`_ can also be
+Getting started with GitHub Codespaces
+----------------------
+
+To get started, create a codespace for this repository by clicking this 👇
```
This came out of some discussion in an astropy dev telecon - we want to make it clear that this isn't a requirement or anything, but just a useful way to get started quickly.

```suggestion
Codespaces is a cloud development environment supported by GitHub.  None of the Astropy build machinery depends on it, but it is a convenient way to quickly get started doing development on Astropy.

To get started, create a codespace for this repository by clicking this 👇
```

eteq reviewed (approved):
I think this is a good idea to provide!  For some added context for anyone else reading this (discussed also in a dev telecon 3-3-23), this isn't really much machinery we as Astropy have to support, but rather just provides a convenience that some developers might like, and definitely could be a useful way to get new contributors up-and-running quickly without much fuss.

pllim commented:
@samruddhikhandale , if you do not mind, please squash all the commits into one commit after you have addressed the review comments. And I think after that, this would be good to go in. Thank you very much!

WilliamJamieson commented:
I just tried the web codespace in this PR and it started with my VScode settings. I think everything worked alright, though I did not poke around too much with it.

This looks really useful for when I need to work away from my normal workstation.

WilliamJamieson commented:
I do have one question, can I open and use this codespace from an iPad? I don't have access at the moment to my iPad to try it out for myself, but it would be useful in a pinch.

samruddhikhandale commented:
> I've noticed each time I make a brand new codespace that I get something like below. It is easily fixed by clicking the button and reloading the editor, but I wonder if there is some configuration issue the is preventing the Python extension from loading initially.

@taldcroft Looks like an upstream VsCode issue, I have contacted the team and reported it.

> Some changes that I think are still needed:

Updated the PR, thank you! 

>  If you do not mind, please squash all the commits into one commit after you have addressed the review comments. And I think after that, this would be good to go in. Thank you very much!

@pllim I have squashed the commits into one, let me know if it looks good. Thank you! 
I am sure you'd be aware of [configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests), it will avoid asking the PR authors to squash commits by themselves. We use this feature on our repos, just thought of sharing. :) 

> I do have one question, can I open and use this codespace from an iPad? I don't have access at the moment to my iPad to try it out for myself, but it would be useful in a pinch.

@WilliamJamieson  Yep, you can open a codespace from an iPad's web browser 🎉

pllim commented:
> I am sure you'd be aware

Haha... we are, however, please see:

* https://github.com/astropy/astropy/issues/14473

pllim commented:
> I have squashed the commits into one

I still see 12 commits. Did you forget to force push?

samruddhikhandale commented:
> I still see 12 commits. Did you forget to force push?

@pllim (Embarrassing nod) Yeah 🤦‍♀️

pllim reviewed (approved):
LGTM. Thanks!

pllim commented:
Since this does touch setup.cfg, I'll let the CI run.

pllim commented:
Thanks, all!

taldcroft commented:
Thanks and amazing job @samruddhikhandale!! This was really a gift from out of the blue since we were talking about codespaces but didn't quite have the expertise or time to dive in. 🎉

colkassad commented:
This is so cool. I want to recreate some of the code I wrote for an astronomy class I took as a way to learn this package. When I created my codespace I couldn't figure out how to plot examples or find any already-integrated notebook functionality. Anyway, I ran 'pip install notebook' which installed Jupyter Notebook to my container, and running 'jupyter notebook' set up a URL I could click to open JN in a new browser tab. This worked and I noticed all the project files are there! However, when I try to create a new notebook, I get "An error occurred while creating a new notebook. Not Found". Am I going about this the wrong way? In the meantime I can surely do this locally but I really like the idea of codespaces as I'm on multiple computers throughout the day. Thanks for providing this!

taldcroft commented:
@colkassad - These days I'm using the VS code notebook instead of `jupyter notebook` in a browser. The VS code integration is great and it pretty much works like you would expect. Use command `Create: New Jupyter Notebook`.

samruddhikhandale commented:
@colkassad Can you follow the update your dev config add install jupyterlab as mentioned in https://docs.github.com/en/codespaces/developing-in-codespaces/getting-started-with-github-codespaces-for-machine-learning#opening-your-codespace-in-jupyterlab?
This should help with `Opening a codespace in Jupyterlab`

colkassad commented:
@samruddhikhandale Thank you, that worked...sort of. I had to add some options to get CORS working properly:

`
"features": {
        "ghcr.io/devcontainers/features/python:1": {
            "installJupyterlab": "true",
            "configureJupyterlabAllowOrigin": "<my codespace URL>"
        }
}
`

Unfortunately matplotlib wouldn't import in a notebook (didn't recognize it as installed) even though "pip install matplotlib" from the terminal reported it was. Trying to !pip install directly in the notebook disconnected the notebook from the kernel. To be clear, I'm using the web version of VSCode that comes up when the codespace boots. Anyway, I don't want to derail this PR discussion anymore. I'll try with a default configuration and add astropy to that and see what happens. Thanks for the help!
